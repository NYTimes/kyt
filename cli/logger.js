
const logger = console;
const write = (status, text, verbose) => {
  let textToLog = '';
  let logObject = false;

  if (status === 'task') textToLog = '👍  ';
  else if (status === 'start') textToLog = '🔥  ';
  else if (status === 'end') textToLog = '✅  ';
  else if (status === 'info') textToLog = 'ℹ️  ';
  else if (status === 'warn') textToLog = '🙀  ';
  else if (status === 'error') textToLog = '❌  ';
  else if (status === 'debug') textToLog = '🐞  ';

  textToLog += text;

  // Adds optional verbose output
  if (verbose) {
    if (typeof verbose === 'object') {
      logObject = true;
    } else {
      textToLog += `\n${verbose}`;
    }
  }

  logger.log(textToLog);
  if (logObject) logger.dir(verbose, { depth: 15 });
};
// Printing any statements
const log = (text) => {
  logger.log(text);
};

// Starting a process
const start = (text) => {
  write('start', text);
};

// Ending a process
const end = (text) => {
  write('end', text);
};

// Tasks within a process
const task = (text) => {
  write('task', text);
};

// Info about a process task
const info = (text) => {
  write('info', text);
};

// Verbose output
// takes optional data
const debug = (text, data) => {
  write('debug', text, data);
};

// Warn output
const warn = (text, data) => {
  write('warn', text, data);
};

// Error output
// takes an optional error
const error = (text, err) => {
  write('error', text, err);
};

module.exports = {
  log,
  task,
  info,
  debug,
  warn,
  error,
  start,
  end,
};
