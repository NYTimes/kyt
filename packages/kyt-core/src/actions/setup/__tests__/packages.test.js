import fs from 'fs';
import shell from 'shelljs';
import logger from 'kyt-utils/logger';
import {
  addKytDependency,
  addPackageJsonScripts,
  checkStarterKytVersion,
  updatePackageJSONDependencies,
  updateUserPackageJSON,
} from '../packages';

jest.mock('fs', () => ({
  writeFileSync: jest.fn(),
  readFileSync: jest.fn(),
}));

jest.mock('shelljs', () => ({
  test: jest.fn(),
  exec: jest.fn(() => ({ code: 0 })),
}));

jest.mock('kyt-utils/logger', () => ({
  info: jest.fn(),
  error: jest.fn(),
  task: jest.fn(),
  warn: jest.fn(),
}));

const userRootPath = '/path/to/repo';
const userPackageJSONPath = `${userRootPath}/package.json`;

describe('packages', () => {
  beforeEach(() => {
    fs.writeFileSync.mockReset();
    fs.readFileSync.mockReset();
    logger.info.mockReset();
    logger.error.mockReset();
    logger.task.mockReset();
    logger.warn.mockReset();
    shell.exec.mockReset();
  });

  describe('addKytDependency', () => {
    test('has dep', () => {
      const result = addKytDependency({ dependencies: { kyt: '*' } });

      expect(result.dependencies.kyt).toEqual('*');
      expect(result.devDependencies).toBeUndefined();
    });

    test('has devDep', () => {
      const result = addKytDependency({ devDependencies: { kyt: '*' } });

      expect(result.devDependencies.kyt).toEqual('*');
      expect(result.dependencies).toBeUndefined();
    });

    test('has no kyt', () => {
      const result = addKytDependency({}, '1.0.0');

      expect(result.devDependencies.kyt).toEqual('1.0.0');
      expect(result.dependencies).toBeUndefined();
    });

    test('has no kyt or preference', () => {
      shell.exec.mockImplementationOnce(() => ({
        stdout: '2.2.2',
      }));

      const result = addKytDependency({});

      expect(result.devDependencies.kyt).toEqual('2.2.2');
      expect(result.dependencies).toBeUndefined();
    });
  });

  describe('addPackageJsonScripts', () => {
    test('no scripts', () => {
      const result = addPackageJsonScripts({});

      expect(result.scripts).toMatchSnapshot();
      expect(logger.task).toHaveBeenCalled();
    });

    test('scripts', () => {
      const result = addPackageJsonScripts({ scripts: { test: 'jest', foo: 'bar' } });

      expect(result.scripts).toMatchSnapshot();
      expect(logger.task).toHaveBeenCalled();
    });

    test('no scripts + temp scripts + deprecated', () => {
      const mockExit = jest.spyOn(process, 'exit').mockImplementation(() => {});

      addPackageJsonScripts({}, { scripts: { foo: 'bar' }, kyt: { scripts: ['foo'] } });

      expect(mockExit).toHaveBeenCalled();
      expect(logger.error).toHaveBeenCalled();
    });

    test('scripts + temp scripts', () => {
      const result = addPackageJsonScripts(
        { scripts: { foo: 'bar' } },
        { scripts: { foo: 'baz' } }
      );

      expect(result.scripts).toEqual(expect.objectContaining({ foo: 'baz' }));
      expect(logger.task).toHaveBeenCalled();
    });
  });

  describe('checkStarterKytVersion', () => {
    test('no preferred dep', () => {
      const result = checkStarterKytVersion({}, { dependencies: {} });

      expect(result).toBeNull();
    });

    test('no preferred devDep', () => {
      const result = checkStarterKytVersion({}, { devDependencies: {} });

      expect(result).toBeNull();
    });

    test('preferred dep', () => {
      const result = checkStarterKytVersion({}, { dependencies: { kyt: '1.0.0' } });

      expect(result).toBe('1.0.0');
    });

    test('preferred devDep', () => {
      const result = checkStarterKytVersion({}, { devDependencies: { kyt: '1.0.0' } });

      expect(result).toBe('1.0.0');
    });

    test('preferred dep + both set', () => {
      const result = checkStarterKytVersion(
        { devDependencies: { kyt: '^1.0.0' } },
        { dependencies: { kyt: '1.0.3' } }
      );

      expect(result).toBe('1.0.3');
      expect(logger.warn).not.toHaveBeenCalled();
    });

    test('preferred devDep + both set', () => {
      const result = checkStarterKytVersion(
        { dependencies: { kyt: '^1.0.0' } },
        { devDependencies: { kyt: '1.0.3' } }
      );

      expect(result).toBe('1.0.3');
      expect(logger.warn).not.toHaveBeenCalled();
    });

    test('preferred dep + both set + warning', () => {
      const result = checkStarterKytVersion(
        { devDependencies: { kyt: '0.0.2' } },
        { dependencies: { kyt: '1.0.0' } }
      );

      expect(result).toBe('1.0.0');
      expect(logger.warn).toHaveBeenCalled();
    });

    test('preferred devDep + both set + warning', () => {
      const result = checkStarterKytVersion(
        { dependencies: { kyt: '0.0.2' } },
        { devDependencies: { kyt: '1.0.0' } }
      );

      expect(result).toBe('1.0.0');
      expect(logger.warn).toHaveBeenCalled();
    });
  });

  describe('updatePackageJSONDependencies', () => {
    test('merge', () => {
      const result = updatePackageJSONDependencies(
        { dependencies: { react: '1' } },
        { dependencies: { react: '2' } }
      );

      expect(result).toEqual({ dependencies: { react: '2' } });
      expect(result.resolutions).toBeUndefined();
    });

    test('duplicates allowed', () => {
      const result = updatePackageJSONDependencies(
        { dependencies: { react: '1' } },
        { devDependencies: { react: '2' } }
      );

      expect(result).toEqual({ dependencies: { react: '1' }, devDependencies: { react: '2' } });
      expect(result.resolutions).toBeUndefined();
    });

    test('resolutions merged', () => {
      const result = updatePackageJSONDependencies(
        { dependencies: { react: '1' } },
        { devDependencies: { react: '2' }, resolutions: { react: '2' } }
      );

      expect(result).toEqual({
        resolutions: { react: '2' },
        dependencies: { react: '1' },
        devDependencies: { react: '2' },
      });
    });
  });

  describe('updateUserPackageJSON', () => {
    test('existing packagejson', () => {
      shell.test.mockImplementationOnce(() => true);
      fs.readFileSync.mockImplementationOnce(() =>
        JSON.stringify(
          {
            name: 'foo',
            version: '3.1.0',
            description: 'a repo',
            main: 'index.js',
            author: 'baz',
            license: 'MIT',
          },
          null,
          2
        )
      );

      const result = updateUserPackageJSON({ userPackageJSONPath }, '1.0.0', {});

      expect(result).toMatchSnapshot();
    });

    test('no existing packagejson', () => {
      shell.test.mockImplementationOnce(() => false);

      const result = updateUserPackageJSON({ userPackageJSONPath }, '1.0.0', {});

      expect(result).toMatchSnapshot();
    });

    test('not existing + no pref', () => {
      shell.test.mockImplementationOnce(() => false);
      shell.exec.mockImplementationOnce(() => ({ stdout: '0.3.0' }));

      const result = updateUserPackageJSON({ userPackageJSONPath }, undefined, {});

      expect(result).toMatchSnapshot();
    });
  });
});
