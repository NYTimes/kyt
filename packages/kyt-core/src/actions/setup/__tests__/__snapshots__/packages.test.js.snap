// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packages addPackageJsonScripts no scripts 1`] = `
Object {
  "build": "kyt build",
  "dev": "kyt dev",
  "kyt:help": "kyt --help",
  "start": "node build/server/main.js",
}
`;

exports[`packages addPackageJsonScripts scripts 1`] = `
Object {
  "build": "kyt build",
  "dev": "kyt dev",
  "foo": "bar",
  "kyt:help": "kyt --help",
  "start": "node build/server/main.js",
  "test": "jest",
}
`;

exports[`packages updateUserPackageJSON existing packagejson 1`] = `
Object {
  "oldPackageJSON": Object {
    "author": "baz",
    "description": "a repo",
    "license": "MIT",
    "main": "index.js",
    "name": "foo",
    "version": "3.1.0",
  },
  "userPackageJSON": Object {
    "author": "baz",
    "description": "a repo",
    "devDependencies": Object {
      "kyt": "1.0.0",
    },
    "license": "MIT",
    "main": "index.js",
    "name": "foo",
    "scripts": Object {
      "build": "kyt build",
      "dev": "kyt dev",
      "kyt:help": "kyt --help",
      "start": "node build/server/main.js",
    },
    "version": "3.1.0",
  },
}
`;

exports[`packages updateUserPackageJSON no existing packagejson 1`] = `
Object {
  "oldPackageJSON": Object {
    "author": "",
    "description": "",
    "license": "",
    "main": "",
    "name": "",
    "version": "1.0.0",
  },
  "userPackageJSON": Object {
    "author": "",
    "description": "",
    "devDependencies": Object {
      "kyt": "1.0.0",
    },
    "license": "",
    "main": "",
    "name": "",
    "scripts": Object {
      "build": "kyt build",
      "dev": "kyt dev",
      "kyt:help": "kyt --help",
      "start": "node build/server/main.js",
    },
    "version": "1.0.0",
  },
}
`;

exports[`packages updateUserPackageJSON not existing + no pref 1`] = `
Object {
  "oldPackageJSON": Object {
    "author": "",
    "description": "",
    "devDependencies": Object {
      "kyt": "1.0.0",
    },
    "license": "",
    "main": "",
    "name": "",
    "scripts": Object {
      "build": "kyt build",
      "dev": "kyt dev",
      "kyt:help": "kyt --help",
      "start": "node build/server/main.js",
    },
    "version": "1.0.0",
  },
  "userPackageJSON": Object {
    "author": "",
    "description": "",
    "devDependencies": Object {
      "kyt": "1.0.0",
    },
    "license": "",
    "main": "",
    "name": "",
    "scripts": Object {
      "build": "kyt build",
      "dev": "kyt dev",
      "kyt:help": "kyt --help",
      "start": "node build/server/main.js",
    },
    "version": "1.0.0",
  },
}
`;
